assets/js/base/hooks/use-collection-data.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * External dependencies
+ */
+import { useState, useEffect, useMemo } from '@wordpress/element';
+import {
+	useQueryStateByContext,
+	useQueryStateByKey,
+	useCollection,
+} from '@woocommerce/base-hooks';
+import { useCollectionDataContext } from '@woocommerce/base-context/collection-data-context';
+import { useDebounce } from 'use-debounce';
+import { find, sortBy } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { useShallowEqual } from './use-shallow-equal';
+
+const buildCollectionDataQuery = ( collectionDataQueryState ) => {
+	const query = collectionDataQueryState;
+
+	if ( collectionDataQueryState.calculate_attribute_counts ) {
+		query.calculate_attribute_counts = sortBy(
+			collectionDataQueryState.calculate_attribute_counts.map(
+				( { taxonomy, queryType } ) => {
+					return {
+						taxonomy,
+						query_type: queryType,
+					};
+				}
+			),
+			[ 'taxonomy', 'query_type' ]
+		);
+	}
+
+	return query;
+};
+
+export const useCollectionData = ( {
+	queryAttribute,
+	queryPrices,
+	queryState,
+	context,
+} ) => {
+	const collectionDataContext = useCollectionDataContext();
+	context = context || collectionDataContext;
+
+	const [ collectionDataQueryState ] = useQueryStateByContext( context );
+	const [
+		calculateAttributesQueryState,
+		setCalculateAttributesQueryState,
+	] = useQueryStateByKey( 'calculate_attribute_counts', [], context );
+	const [
+		calculatePriceRangeQueryState,
+		setCalculatePriceRangeQueryState,
+	] = useQueryStateByKey( 'calculate_price_range', false, context );
+
+	const currentQueryAttribute = useShallowEqual( queryAttribute || {} );
+	const currentQueryPrices = useShallowEqual( queryPrices );
+
+	useEffect( () => {
+		if ( currentQueryAttribute ) {
+			const foundAttribute = find(
+				calculateAttributesQueryState,
+				( attribute ) => {
+					return (
+						attribute.taxonomy === currentQueryAttribute.taxonomy
+					);
+				}
+			);
+
+			if ( ! foundAttribute ) {
+				const setState = calculateAttributesQueryState;
+				setState.push( currentQueryAttribute );
+				setCalculateAttributesQueryState( setState );
+			}
+		}
+	}, [ currentQueryAttribute ] );
+
+	useEffect( () => {
+		if (
+			calculatePriceRangeQueryState !== currentQueryPrices &&
+			currentQueryPrices !== undefined
+		) {
+			setCalculatePriceRangeQueryState( currentQueryPrices );
+		}
+	}, [ currentQueryPrices ] );
+
+	// Defer the select query so all collection-data query vars can be gathered.
+	const [ shouldSelect, setShouldSelect ] = useState( false );
+	const [ debouncedShouldSelect ] = useDebounce( shouldSelect, 200 );
+
+	if ( ! shouldSelect ) {
+		setShouldSelect( true );
+	}
```

Comment:
This is a bit confusing to me.  

First render: `shouldSelect = false`, `debouncedShouldSelect=false` - `setShouldSelect` fires
Second render: `shouldSelect=true`, debouncedShouldSelect=false`( if 200ms hasn't elapsed) - `setShouldSelect` does not fire.
Third render (after 200ms): `shouldSelect=true`, `debouncedShouldSelect=true` - `setShouldSelect` does not fire.
...all future renders the same values as above ^^ are returned.  This _only_ resets when a component using this hook unmounts.

If I'm reading this correctly, this just delays collection data requests on initial mount for 200ms which if that's the goal seems okay.  So I wrote all this just so you can confirm I'm understanding things correctly :)

Reply:
Exactly that. The first render is where the problems occur because all the filter blocks are rendering one by one and doing queries. By deferring this 200ms (roughly) they all have a chance to use this hook so a single query can run.